### PR TITLE
Fix bugs that can result in endless iterations

### PR DESCRIPTION
**Pull request by bertfrees**
There are three main commits in this PR that fix three different bugs, all related to the CrossReferenceHandler.

- 4d88f38: Uses BlockLineLocation instead of SheetIdentity to store "breakable" info, for the same reason why we're using BlockLineLocation for the "volume transition properties" info (since commit d386effaf64890cf587911ebdd73).
- a98cc81: Fixes a bug that could cause endless iterations when the last sheet of a volume had a vertically positioned block on it.
- a55be8b: Fixes a bug that could cause endless iterations when the volume was split at a sheet with an empty back.

The first and the last I discovered the hard way, because Dotify would fail to convert certain books. The second one is something I discovered in the code while looking for the two other bugs.

The 4 other commits are small refactorings are fixes to comments.

@kalaspuffar @PaulRambags Please have a look. Hope you can make sense of it. I've tried to explain well in the code why certains things are done, but let me know if anything needs more clarification.

### CODE REVIEW

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

Could you move this PR to the dotify.library repository instead?

Best regards
Daniel

**Comment by bertfrees** [comment]
OK

**Review by PaulRambags — COMMENTED** [review]
It would be nice to have a unit test which would fail before your improvements and which passes now. 
I can imagine, though, that this is hard when it comes to endless iterations.

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:310]: Please fix the checkstyle errors.

`/home/travis/build/mtmse/dotify.formatter.impl/src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:24:8: Unused import`

`import org.daisy.dotify.formatter.impl.core.TransitionContent.Type;` should be removed

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:303]: Please align this comment correctly

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:312]: Is this also valid in simplex mode?

File: src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java:33]: I would expect here:

`this.blockLineLocation = blockLineLocation;`

File: src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java:49]: Ok it is here... at the end. Syntactically correct but usually you put things in the same order as the parameters

File: src/org/daisy/dotify/formatter/impl/sheet/SheetDataSource.java
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/sheet/SheetDataSource.java:418]: I'd prefer a more descriptive name than just `n` so that it's clear what `n` stands for
Or, add Javadoc

**Review by PaulRambags — CHANGES_REQUESTED** [review]

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:312]: Yes. In simplex mode nextEmpty will never be true. I'll change the comment a bit.

File: src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java:49]: Really?

File: src/org/daisy/dotify/formatter/impl/sheet/SheetDataSource.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/sheet/SheetDataSource.java:418]: I'll add a Javadoc.

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:303]: What is the correct alignment?

**Comment by bertfrees** [comment]
> It would be nice to have a unit test which would fail before your improvements and which passes now.
> I can imagine, though, that this is hard when it comes to endless iterations.

Yeah I know. It's kinda hard and probably time consuming to come up with tests for fixes like these. In some cases the bug is revealed only by a specific series of events that are sometimes encountered when transforming real books, but that are not so easily reproducible. Moreover, I'm not so sure the tests that I would come up with would be very useful as regression tests because whether or not the algorithm ends up in an endless loop depends on several other things.

I can give it a shot but it might take a while.

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (PaulRambags) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:303]: I would say, give it the same indent as the next line (which starts with `if`) so that the comment aligns better. Now it is very much to the right.
But this is a minor detail, don't spent too much time on it

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:303]: I've intentionally aligned it with the comment on the previous line because it's one and the same comment. It got messed up when the whole code base was reformatted.

**Comment by bertfrees** [comment]
> Could you move this PR to the dotify.library repository instead?

Actually I was waiting for a release of dotify.library before switching to it in DAISY Pipeline. We said we were gonna make a release of dotify.library that is equivalent to the latest releases of the separate dotify projects. My "[new-merge](https://github.com/bertfrees/dotify.library/)" branch also still hasn't been integrated yet.

**Comment by bertfrees** [comment]
@kalaspuffar Moved to the dotify.library repo: https://github.com/bertfrees/dotify.library/compare/new-merge...bertfrees:fix-endless-iterations

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

Ok, not seen a PR for these changes. Should we do a release with the current dotify.library repo or do I still wait for your PR with all the commit history?

Best regards
Daniel

**Comment by bertfrees** [comment]
Do you need a PR? You just need to reset the main branch to my branch.

**Comment by bertfrees** [comment]
Just tried it, Github says I can't make a PR when the branches have entirely different commit histories.

> Should we do a release with the current dotify.library repo or do I still wait for your PR with all the commit history?

Wait for all the commit history.

There are also some issues I'm discovering now which I think should be fixed before making a release.
- There are some packages that should be exported (add a packageinfo file):
  - org.daisy.braille.utils.pef
  - org.daisy.dotify.common.io
  - org.daisy.dotify.common.text
  - org.daisy.dotify.common.xml
- and there is also a package that was omitted, namely org.daisy.braille.utils.impl.tasks (presumably because it depends on streamline-api)
- There are also some file missing from META-INF/services, but that was already an issue before the merge.

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

I'll wait until you have looked into these issues as you know what needs to be exported for instance.

The task package was not added as it's not used anymore. Both task packages where required by Dotify Studio and aren't needed in the new library. If we need those packages in the future we will add them to the Studio code instead.

I think we have discussed that issue on a meeting but I might be miss remembering that fact.

What file was missing ? Just curious? I thought I remembered to merge all files in all repositories.

Best regards
Daniel

**Comment by bertfrees** [comment]
> I think we have discussed that issue on a meeting but I might be miss remembering that fact.

Possible, but I don't remember.

I noticed it because we expose the dotify.tasks.impl through a XProc step in Pipeline, and in order for it to work, the org.daisy.braille.utils.impl.tasks package is needed. This functionality is not actually used in Pipeline though, so if you omitted it on purpose and you have no intention to maintain the project, I can just remove it from Pipeline too.

The file that is missing from META-INF/services is:

- org.daisy.dotify.api.embosser.FileFormatProvider

       org.daisy.braille.utils.impl.provider.BrailleEditorsFileFormatProvider

I've worked around it by simply adding them in one of the Pipeline modules. That works too, even though the classes are not in the same JAR.

Again, this is not a new issue, so maybe it's better to do it after the first releases.

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:310]: Fixed in https://github.com/bertfrees/dotify.library/commit/137db797379da574c1ebf51aee1189771c63f28d.

File: src/org/daisy/dotify/formatter/impl/sheet/SheetDataSource.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/sheet/SheetDataSource.java:418]: @PaulRambags See https://github.com/bertfrees/dotify.library/commit/bb605b079edf38b6eacb3db172792ef701bb697c.

File: src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:312]: See https://github.com/bertfrees/dotify.library/commit/0c95dc06ae7c21fa536560254e3d10df86818cf7

**Comment by bertfrees** [comment]
I've created a new PR in the dotify.library repository: https://github.com/mtmse/dotify.library/pull/9

**Pull request closed without merging**
